pkg/reconcilermanager/controllers/errors.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"fmt"
+
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	kstatus "sigs.k8s.io/cli-utils/pkg/kstatus/status"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// OperationCreate is the create operation
+	OperationCreate = Operation("create")
+	// OperationUpdate is the update operation
+	OperationUpdate = Operation("update")
+	// OperationPatch is the patch operation
+	OperationPatch = Operation("patch")
+	// OperationDelete is the delete operation
+	OperationDelete = Operation("delete")
+	// OperationGet is the get operation
+	OperationGet = Operation("get")
+	// OperationList is the list operation
+	OperationList = Operation("list")
+	// OperationWatch is the watch operation
+	OperationWatch = Operation("watch")
+)
+
+// Operation performed on a Kubernetes resource or object
+type Operation string
+
+// ObjectOperationError is an error from the reconciler-manager regarding
+// failure to perform an operation on a managed Kubernetes resource or resource
+// object.
+type ObjectOperationError struct {
+	// ID of the managed object
+	ID core.ID
+	// Operation attempted on the managed object
+	Operation Operation
+	// Cause of the operation failure
+	Cause error
+}
+
+// Error returns the error string
+func (ooe *ObjectOperationError) Error() string {
+	return fmt.Sprintf("resource object operation failed (operation: %q, object: %q, objectKind: %q): %v",
+		ooe.Operation, ooe.ID.ObjectKey, ooe.ID.Kind, ooe.Cause)
+}
+
+// NewObjectOperationError constructs a new ObjectOperationError
+func NewObjectOperationError(err error, obj client.Object, op Operation) *ObjectOperationError {
+	id := core.IDOf(obj)
+	if id.GroupKind.Empty() {
+		gvk, lookupErr := kinds.Lookup(obj, core.Scheme)
+		if lookupErr != nil {
+			// Invalid or unregistered resource type - err probably already includes the same message
+			// Fake the kind with the struct type
+			id.Kind = fmt.Sprintf("%T", obj)
+		}
+		id.GroupKind = gvk.GroupKind()
```

Comment:
Is `gvk` valid if the `lookupErr` is non-nil? I expect perhaps we want this to be in an `else`

Reply:
fixed